### ISSUE TITLE
As a User, I want to view the top viewed and downloaded resources and published resources on HydroShare aggregated over some timescale

### ISSUE THREAD

**Issue by aaraney** [hydroshare/hydroshare#4478]
## Description

@ClaraCogswell and I want a view (page or inset in existing page) that lists the top `n` viewed and top `n` downloaded resources over the past event timescale. A filter should be present for selecting to view publish resources or (inclusive or) published resources. The view likely would take the form of two tables or one table with two columns. At the top of the table, a user could select from a drop down box the timescale of interest (e.g. day, week, month, year). Items in the table would consist of the resource's title and primary author. Additional a defined set of flares (i.e. 
![image](https://user-images.githubusercontent.com/39420640/153069164-273ce330-4082-4fa4-96fb-f70fc7e5f427.png)) could be added beneath resources in the list to denote for example, if a resource contains code, is published, or is considered reproducible. 

| top viewed | top downloaded |
| ----------- | ---------------- |
| [Notebook for retrieval of National Water Model Retrospective run results at SNOTEL sites](https://www.hydroshare.org/resource/3d4976bf6eb84dfbbe11446ab0e31a0a/) | [Presentations for iRODS User Group meeting](https://www.hydroshare.org/resource/53bab1bbcd544ad78e45443da7246fe4/) |
| ... | ... |
| ... | ... |
| ... | ... |
| ... | ... |

However, for scoping purposes, the feature I am describing above would not let the user "traverse time." Meaning, the user could not select an arbitrary date range and populate the aforementioned tables.

## Why

This feature could enrich the discovery experience and highlight work that is gaining traction on HydroShare. This feature also has the potential to add value to HydroShare. Not only is HydroShare a community but it is also a platform to become better known within the community. This view would be quickly digestible and approachable as it would be similar to features present on platforms like youtube or reddit.

**Comment by ClaraCogswell** [comment]
We think this feature addition would have several benefits, the feature would:
Display, by proxy, which resources the community is currently most interested in.
Highlight high impact content stored in HS.
Exhibit community engagement via HS.
Implicitly showcase a benefit to publishing content on HS (i.e. free community advertising of resources that are receiving attention).
Enhance the HS resource discovery experience.
We propose the feature takes the following form:

A table that displays the top 5 viewed published HS resources for the previous month in ascending ranked order. Here, a view is qualified when a user clicks on a HS minted DOI and their click successfully resolves back to the DOI's HS resource. The table would include resources' rank, title (should be a link to the resource), author(s), and publication date. A month date selector would be included to allow a user to “go back in time” and browse, at the month scale, to see which resources were most viewed during their specified period.

The data necessary for this feature is be provided by a monthly report generated by CrossRef. The monthly report lists the number of successful DOI resolutions (successful redirect when clicking a DOI to arriving at a HydroShare resource) for CUAHSI’s DOI prefix. As such, it is a seemingly good candidate to be developed as a “micro-service” that HS could embed within relevant pages (HS home page, discover, etc.).

The inspiration for this project came from crossref's monthly report that includes the top 10 most visited DOIs for the past 30 days. This feature could be a potentially valuable way to utilize this data. Of course, from you we are looking for constructive feedback and perspectives on the value of this effort. We'll be able to provide more materials and overview when we meet.

**Comment by aaraney** [comment]
This feature could be integrated into the hydroshare ecosystem without introducing a large amount of code into the code base. The diagram below shows, at a high level, how this feature could fit in the existing ecosystem.

```mermaid
flowchart LR

subgraph nginx
c[proxy]
d[proxy]
end
c--> a[HydroShare django application]
d --> b[top published resources service]
```

In this design, the top published resources service would be placed behind HydroShare's nginx http server via proxy. This would allow HydroShare, the website, to treat both the django app and the top published resources service as if they were on the same server. Minimal javascript code would be added to HydroShare pages that want to consume data from the top published resources service. Code to display the information however would also be added via javascript.

**Comment by horsburgh** [comment]
I think this functionality could be appropriate for the Home page, but I would vote for it not getting added to the My Resources page - or at least not as mocked up above. 

I interact a lot with the My Resources page, and I don't want to have to wade through a section at the the top that I would rarely use to get to the actual functionality of the page. If we added functionality like this to the My Resources page, maybe add it as a filter on the left - e.g., filter to show me my top viewed resources or filter to show me my top downloaded resources - and the table would be filtered or sorted to list resources in order. Similar filters/sorting could be added to the Discover page. The layout and function of the My Resources and Discover pages would be essentially unchanged.

For the Home page, I think we could come up with some better styling to more effectively use space on the page. Could that section be a little more interactive? Change the section header to "Top Resources" and then give the user a few set choices - e.g., show me the top downloaded resources, the top cited resources, resources with the most views, etc. You could add some simple set time filters too - e.g., over all time, this year, in the last month. Then present the resources in a nice table that takes the full width of the page as the Recently Visited table does.

Instead of modifying the My Resources page, this could also be done on a user's Profile page where we have an Overview of the user's resources and a listing of their contributed resources. In the overview section we could list Recent Activity and Top Resources.

**Comment by aaraney** [comment]
@horsburgh, I should have noted that the mock-ups are purely proofs of concept; They are by no means set in stone. We hoped they could be used to get others on the same page when it came to this feature addition. I agree that we do not want to crowd the "My Resources" page. What are your thoughts on including the feature on the "My Resources" page in a collapsible panel that is by default collapsed? So, take the table in the above mockup and put it in a collapsible element.

> For the Home page, I think we could come up with some better styling to more effectively use space on the page. 

I completely agree. I hope that as we get others on board with the idea, so too will stylistic opinions. 

>  Could that section be a little more interactive? Change the section header to "Top Resources" and then give the user a few set choices - e.g., show me the top downloaded resources, the top cited resources, resources with the most views, etc. You could add some simple set time filters too - e.g., over all time, this year, in the last month. Then present the resources in a nice table that takes the full width of the page as the Recently Visited table does.

I love this idea. Originally, I wanted to move in this direction to a more holistic summary view. However, there were some internal concerns about the meaning of _top downloaded _ and _top viewed_ in regard to unpublished resources. I think others were concerns that given the way these "metrics" are collected by HydroShare the software. They might not have meaning or would be hard to interpret. @Castronova and @martinseul can you all elaborate on your concerns in regard to @horsburgh's above quote?

> Instead of modifying the My Resources page, this could also be done on a user's Profile page where we have an Overview of the user's resources and a listing of their contributed resources. In the overview section we could list Recent Activity and Top Resources.

I agree. I think the scope of "where" to put this feature on HydroShare is still very much up for debate.

**Comment by martinseul** [comment]
I agree with Jeff sentiment and also see the issue of adding additional elements to the current existing pages. This idea fits more into a concept we had discussed before and that we called "browsing" instead of "discovery", the use case for discovery is I'm searching for something specific while browsing is more show me what you have sorted by some form of criteria e.g. most downloaded or most relevant like a recommendation system.
